lib/spack/llnl/util/link_tree.py

```diff
@@ -112,4 +126,6 @@ def unmerge(self, dest_root, **kwargs):
             elif os.path.exists(dest):
                 if not os.path.islink(dest):
                     raise ValueError("%s is not a link tree!" % dest)
-                os.remove(dest)
+                # only remove if src actually points to dest
+                if filecmp.cmp(src, dest, shallow=True):
```

Comment:
Using `realpath` would be sufficient if we only ever used symlinks, but for hardlinks the `realpath`s differ.

Although I do admit that this should only ever be an issue when adding (and subsequently removing) packages containing conflicting files with hardlinks. In all other cases the original `os.remove` would have been sufficient.